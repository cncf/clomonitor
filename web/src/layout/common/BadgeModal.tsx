import CodeBlock from './CodeBlock';
import Modal from './Modal';
import Tabs from './Tabs';

interface OpenModalStatus {
  status: boolean;
  name?: string;
}

interface Props {
  orgName: string;
  projectName: string;
  openStatus: OpenModalStatus;
  onCloseModal: () => void;
}

const BadgeModal = (props: Props) => {
  const origin = window.location.origin;
  const badgeImage = `https://img.shields.io/endpoint?url=${origin}/api/projects/${props.orgName}/${props.projectName}/badge`;
  const markdownLink = `[![CLOMonitor](${badgeImage})](${origin}/projects/${props.orgName}/${props.projectName})`;
  const asciiLink = `${origin}/projects/${props.orgName}/${props.projectName}[image:${badgeImage}[CLOMonitor]]`;

  return (
    <Modal
      header="Project badge"
      onClose={props.onCloseModal}
      open={props.openStatus.status && props.openStatus.name === 'badge'}
    >
      <div className="my-3">
        <Tabs
          tabs={[
            {
              name: 'markdown',
              title: 'Markdown',
              content: (
                <>
                  <div className="mt-2 mb-4">
                    <img src={badgeImage} alt="CLOMonitor badge" />
                  </div>

                  <CodeBlock language="markdown" content={markdownLink} label="Copy badge markdown link to clipboard" />
                </>
              ),
            },
            {
              name: 'ascii',
              title: 'AsciiDoc',
              content: (
                <>
                  <div className="mt-2 mb-4">
                    <img src={badgeImage} alt="CLOMonitor badge" />
                  </div>

                  <CodeBlock language="asciidoc" content={asciiLink} label="Copy badge Ascii link to clipboard" />
                </>
              ),
            },
          ]}
          active="markdown"
          noDataContent="Sorry, the information for this is missing."
        />
      </div>
    </Modal>
  );
};

export default BadgeModal;
