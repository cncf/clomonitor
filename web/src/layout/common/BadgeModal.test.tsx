import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import BadgeModal from './BadgeModal';

const mockOnCloseModal = jest.fn();

const defaultProps = {
  orgName: 'org',
  projectName: 'proj',
  openStatus: { status: true, name: 'badge' },
  onCloseModal: mockOnCloseModal,
};

describe('BadgeModal', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<BadgeModal {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders markdown tab', () => {
      render(<BadgeModal {...defaultProps} />);

      expect(screen.getByText('Project badge')).toBeInTheDocument();
      expect(screen.getAllByText('Markdown')).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: /Open tab/ })).toHaveLength(2);

      const badge = screen.getByAltText('CLOMonitor badge');
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveProperty(
        'src',
        'https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge'
      );
      const code = screen.getByTestId('code');
      expect(code).toBeInTheDocument();
      expect(code).toHaveTextContent(
        '[![CLOMonitor](https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge)](http://localhost/projects/org/proj)'
      );
    });

    it('renders ascii tab', () => {
      render(<BadgeModal {...defaultProps} />);

      expect(screen.getAllByText('AsciiDoc')).toHaveLength(2);
      const btns = screen.getAllByRole('button', { name: /Open tab/ });
      expect(btns[1]).toHaveTextContent('AsciiDoc');
      userEvent.click(btns[1]);

      const badge = screen.getByAltText('CLOMonitor badge');
      expect(badge).toBeInTheDocument();
      expect(badge).toHaveProperty(
        'src',
        'https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge'
      );
      const code = screen.getByTestId('code');
      expect(code).toBeInTheDocument();
      expect(code).toHaveTextContent(
        'http://localhost/projects/org/proj[image:https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge[CLOMonitor]]'
      );
    });
  });
});
