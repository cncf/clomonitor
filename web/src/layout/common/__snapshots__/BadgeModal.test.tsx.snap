// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeModal creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="modal-backdrop activeBackdrop"
      data-testid="modalBackdrop"
    />
    <div
      aria-modal="true"
      class="modal modal active d-block"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content rounded-0 border border-3 mx-auto content"
        >
          <div
            class="modal-header rounded-0 d-flex flex-row align-items-center header"
          >
            <div
              class="modal-title h5 m-2 flex-grow-1"
            >
              Project badge
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          >
            <div
              class="my-3"
            >
              <div>
                <ul
                  class="nav nav-tabs tabs"
                >
                  <li
                    class="nav-item"
                  >
                    <button
                      aria-label="Open tab markdown"
                      class="btn nav-item rounded-0 lightText btn active btn-secondary active"
                    >
                      <span
                        class="d-none d-sm-block"
                      >
                        Markdown
                      </span>
                      <span
                        class="d-block d-sm-none"
                      >
                        Markdown
                      </span>
                    </button>
                  </li>
                  <li
                    class="nav-item"
                  >
                    <button
                      aria-label="Open tab ascii"
                      class="btn nav-item rounded-0 lightText btn"
                    >
                      <span
                        class="d-none d-sm-block"
                      >
                        AsciiDoc
                      </span>
                      <span
                        class="d-block d-sm-none"
                      >
                        AsciiDoc
                      </span>
                    </button>
                  </li>
                </ul>
              </div>
              <div
                class="tab-content mt-4"
              >
                <div
                  class="tab-pane fade show active"
                >
                  <div
                    class="mt-2 mb-4"
                  >
                    <img
                      alt="CLOMonitor badge"
                      src="https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge"
                    />
                  </div>
                  <div
                    class="d-flex flex-row align-items-center pb-2 codeBlock"
                    data-testid="code"
                  >
                    <pre
                      style="display: block; overflow-x: auto; padding: 1rem 0.5rem; background: rgb(248, 248, 255); margin-bottom: 0px;"
                    >
                      <code
                        class="language-markdown"
                        style="white-space: pre;"
                      >
                        <span>
                          [
                        </span>
                        <span
                          style="color: rgb(33, 145, 97);"
                        >
                          ![CLOMonitor
                        </span>
                        <span>
                          ](
                        </span>
                        <span
                          style="color: rgb(187, 102, 136);"
                        >
                          https://img.shields.io/endpoint?url=http://localhost/api/projects/org/proj/badge
                        </span>
                        <span>
                          )](
                        </span>
                        <span
                          style="color: rgb(187, 102, 136);"
                        >
                          http://localhost/projects/org/proj
                        </span>
                        <span>
                          )
                        </span>
                      </code>
                    </pre>
                    <div
                      class="position-relative ms-3"
                    >
                      <button
                        aria-label="Copy badge markdown link to clipboard"
                        class="btn btn-sm btn-primary rounded-circle btn"
                        type="button"
                      >
                        <div
                          aria-hidden="true"
                          class="d-flex flex-row align-items-center"
                        >
                          <svg
                            fill="none"
                            height="1em"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <rect
                              height="13"
                              rx="2"
                              ry="2"
                              width="13"
                              x="9"
                              y="9"
                            />
                            <path
                              d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                            />
                          </svg>
                        </div>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="modal-footer p-3"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm rounded-0 btn-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
